source/extensions/filters/http/transform/transform.cc

```diff
@@ -0,0 +1,387 @@
+                                 Server::Configuration::ServerFactoryContext& context,
+                                 absl::Status& creation_status)
+    : clear_route_cache_(config.clear_route_cache()),
+      clear_cluster_cache_(config.clear_cluster_cache()) {
+  if (config.has_request_transformation()) {
+    request_transformation_.emplace(config.request_transformation(), context, creation_status);
+  }
+  if (config.has_response_transformation()) {
+    response_transformation_.emplace(config.response_transformation(), context, creation_status);
+  }
+
+  if (clear_cluster_cache_ && clear_route_cache_) {
+    creation_status = absl::InvalidArgumentError(
+        "Only one of clear_cluster_cache and clear_route_cache can be set to true");
+  }
+}
+
+void TransformFilter::maybeInitializeRouteConfigs(Http::StreamFilterCallbacks* callbacks) {
+  // Ensure that route configs are initialized only once and the same route configs are used
+  // for both decoding and encoding paths.
+  // An independent flag is used to ensure even at the case where the route configs is empty,
+  // we still won't try to initialize it again.
+  if (route_configs_initialized_) {
+    return;
+  }
+  route_configs_initialized_ = true;
+
+  // Traverse through all route configs to retrieve all available header mutations.
+  // `getAllPerFilterConfig` returns in ascending order of specificity (i.e., route table
+  // first, then virtual host, then per route).
+  auto route_config = Http::Utility::resolveMostSpecificPerFilterConfig<TransformConfig>(callbacks);
+  if (route_config != nullptr) {
+    effective_config_ = route_config;
+  } else {
+    effective_config_ = config_.get();
+  }
+}
+
+Http::FilterHeadersStatus TransformFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                         bool end_stream) {
+  // Skip transformation for headers only requests or non-JSON requests.
+  if (end_stream || !absl::StrContains(headers.getContentTypeValue(),
+                                       Http::Headers::get().ContentTypeValues.Json)) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // Initialize effective route configs if not done yet.
+  maybeInitializeRouteConfigs(decoder_callbacks_);
+  ASSERT(effective_config_ != nullptr);
+
+  if (!effective_config_->requestTransformation().has_value()) {
+    // No request transform configured, continue.
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // No request transform configured, continue.
+  decoding_enabled_ = true;
+  return Http::FilterHeadersStatus::StopIteration;
+}
+
+Http::FilterDataStatus TransformFilter::decodeData(Buffer::Instance& data, bool end_stream) {
+  if (!decoding_enabled_) {
+    return Http::FilterDataStatus::Continue;
+  }
+  if (end_stream) {
+    decoder_callbacks_->addDecodedData(data, true);
+    handleCompleteRequestBody();
+    return Http::FilterDataStatus::Continue;
+  }
```

Comment:
Related but not tied / blocker to this PR, do you have some insights for the streaming JSON parser support without waiting for the whole body once we get the enough data. I think the current nlohmann lib doesn't quite support it, and we could need some SAX style approach.